max-encoded-len/derive/tools/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// tag::description[]
+//! Proc macro helpers for procedural macros
+// end::description[]
+
+// reexport proc macros
+pub use test_procedural_tools_derive::*;
+
+use proc_macro_crate::{crate_name, FoundCrate};
+use syn::parse::Error;
+use quote::quote;
+
+pub mod syn_ext;
+
+// FIXME #1569, remove the following functions, which are copied from sp-api-macros
+use proc_macro2::{TokenStream, Span};
+use syn::Ident;
+
+fn generate_hidden_includes_mod_name(unique_id: &str) -> Ident {
+	Ident::new(&format!("sp_api_hidden_includes_{}", unique_id), Span::call_site())
+}
+
+/// Generates the access to the `frame-support` crate.
+pub fn generate_crate_access(unique_id: &str, def_crate: &str) -> TokenStream {
+	if std::env::var("CARGO_PKG_NAME").unwrap() == def_crate {
+		quote::quote!( frame_support )
+	} else {
+		let mod_name = generate_hidden_includes_mod_name(unique_id);
+		quote::quote!( self::#mod_name::hidden_include )
+	}
+}
+
+/// Generate the crate access for the crate using 2018 syntax.
+///
+/// for `frame-support` output will for example be `frame_support`.
+pub fn generate_crate_access_2018(def_crate: &str) -> Result<syn::Ident, Error> {
```

Comment:
We could probably move this function directly into `max-encoded-len/derive/src/lib.rs` and then rm both the `temp-procedural-tools` and `test-procedural-tools-derive` crates entirely.

Reply:
I mean sure, but even so it would be a bit embarrassing to accidentally move two entire extra crates into `parity-scale-codec` when we really only need these 15 lines of code.